File: src/main/java/innexgo/hours/AdminshipRequestService.java

/*
 * Innexgo Website
 * Copyright (C) 2020 Innexgo LLC
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package innexgo.hours;

import java.util.List;
import java.util.stream.Stream;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.jdbc.core.RowMapper;
import org.springframework.stereotype.Repository;
import org.springframework.transaction.annotation.Transactional;

@Transactional
@Repository
public class AdminshipRequestService {

  @Autowired private JdbcTemplate jdbcTemplate;

  public AdminshipRequest getByAdminshipRequestId(long adminshipRequestId) {
    String sql =
        "SELECT * FROM adminship_request WHERE adminship_request_id=?";
    RowMapper<AdminshipRequest> rowMapper = new AdminshipRequestRowMapper();
    AdminshipRequest apptRequest = jdbcTemplate.queryForObject(sql, rowMapper, adminshipRequestId);
    return apptRequest;
  }

  public long nextId() {
    String sql = "SELECT max(adminship_request_id) FROM adminship_request";
    Long maxId = jdbcTemplate.queryForObject(sql, Long.class);
    if(maxId == null) {
      return 0;
    } else {
      return maxId + 1;
    }
  }

  public void add(AdminshipRequest apptRequest) {
    apptRequest.adminshipRequestId = nextId();
    apptRequest.creationTime = System.currentTimeMillis();
    // Add apptRequest
    String sql =
        "INSERT INTO adminship_request values (?, ?, ?, ?, ?)";
    jdbcTemplate.update(
        sql,
        apptRequest.adminshipRequestId,
        apptRequest.creationTime,
        apptRequest.creatorUserId,
        apptRequest.schoolId,
        apptRequest.message
    );
  }

  public boolean existsByAdminshipRequestId(long adminshipRequestId) {
    String sql = "SELECT count(*) FROM adminship_request sr WHERE sr.adminship_request_id=?";
    int count = jdbcTemplate.queryForObject(sql, Integer.class, adminshipRequestId);
    return count != 0;
  }

  // Restrict apptRequests by
  public Stream<AdminshipRequest> query(
      Long adminshipRequestId,
      Long creationTime,
      Long minCreationTime,
      Long maxCreationTime,
      Long creatorUserId,
      Long schoolId,
      String message,
      Boolean responded,
      long offset,
      long count) {
    String sql =
        "SELECT sr.* FROM adminship_request sr"
            + (responded       == null ? "" : " LEFT JOIN adminship_request_response srr ON srr.adminship_request_id = sr.adminship_request_id")
            + " WHERE 1=1 "
            + (responded       == null ? "" : " AND srr.adminship_request_id IS" + (responded ? " NOT NULL" : " NULL"))
            + (adminshipRequestId== null ? "" : " AND sr.adminship_request_id = " + adminshipRequestId)
            + (creatorUserId   == null ? "" : " AND sr.creator_user_id = " + creatorUserId)
            + (creationTime    == null ? "" : " AND sr.creation_time = " + creationTime)
            + (minCreationTime == null ? "" : " AND sr.creation_time > " + minCreationTime)
            + (maxCreationTime == null ? "" : " AND sr.creation_time < " + maxCreationTime)
            + (schoolId        == null ? "" : " AND sr.school_id = " + schoolId)
            + (message         == null ? "" : " AND sr.message = " + Utils.escape(message))
            + (" ORDER BY sr.adminship_request_id")
            + (" LIMIT " + offset + ", " + count)
            + ";";

    RowMapper<AdminshipRequest> rowMapper = new AdminshipRequestRowMapper();
    return this.jdbcTemplate.query(sql, rowMapper).stream();
  }
}

File: rust/src/adminship_request_service.rs
use super::db_types::*;
use super::utils::current_time_millis;
use innexgo_hours_api::request;
use tokio_postgres::GenericClient;

impl From<tokio_postgres::row::Row> for AdminshipRequest {
  // select * from adminship_request order only, otherwise it will fail
  fn from(row: tokio_postgres::Row) -> AdminshipRequest {
    AdminshipRequest {
      adminship_request_id: row.get("adminship_request_id"),
      creation_time: row.get("creation_time"),
      creator_user_id: row.get("creator_user_id"),
      school_id: row.get("school_id"),
      message: row.get("message"),
    }
  }
}

pub async fn add(
  con: &mut impl GenericClient,
  creator_user_id: i64,
  school_id: i64,
  message: String,
) -> Result<AdminshipRequest, tokio_postgres::Error> {
  let creation_time = current_time_millis();

  let adminship_request_id = con
    .query_one(
      "INSERT INTO
       adminship_request(
           creation_time,
           creator_user_id,
           school_id,
           message
       )
       VALUES($1, $2, $3, $4)
       RETURNING adminship_request_id
      ",
      &[&creation_time, &creator_user_id, &school_id, &message],
    )
    .await?
    .get(0);

  // return adminship_request
  Ok(AdminshipRequest {
    adminship_request_id,
    creation_time,
    creator_user_id,
    school_id,
    message,
  })
}

pub async fn get_by_adminship_request_id(
  con: &mut impl GenericClient,
  adminship_request_id: i64,
) -> Result<Option<AdminshipRequest>, tokio_postgres::Error> {
  let result = con
    .query_opt(
      "SELECT * FROM adminship_request WHERE adminship_request_id=$1",
      &[&adminship_request_id],
    )
    .await?
    .map(|x| x.into());

  Ok(result)
}

pub async fn query(
  con: &mut impl GenericClient,
  props: request::AdminshipRequestViewProps,
) -> Result<Vec<AdminshipRequest>, tokio_postgres::Error> {
  let sql = "SELECT ar.* FROM adminship_request ar WHERE 1 = 1
     AND ($1::bigint[] IS NULL OR ar.adminship_request_id IN $1)
     AND ($2::bigint   IS NULL OR ar.creation_time >= $2)
     AND ($3::bigint   IS NULL OR ar.creation_time <= $3)
     AND ($4::bigint   IS NULL OR ar.creator_user_id = $4)
     AND ($5::bigint   IS NULL OR ar.school_id = $5)
     AND ($6::text     IS NULL OR ar.message = $6)
     AND ($7::text     IS NULL OR ar.message LIKE CONCAT('%',$7,'%'))
     ORDER BY ar.adminship_request_id
     LIMIT $8
     OFFSET $9
     ";

  let stmnt = con.prepare(sql).await?;

  let results = con
    .query(
      &stmnt,
      &[
        &props.adminship_request_id,
        &props.min_creation_time,
        &props.max_creation_time,
        &props.creator_user_id,
        &props.school_id,
        &props.message,
        &props.partial_message,
        &props.count.unwrap_or(100),
        &props.offset.unwrap_or(0),
      ],
    )
    .await?
    .into_iter()
    .map(|x| x.into())
    .collect();

  Ok(results)
}
